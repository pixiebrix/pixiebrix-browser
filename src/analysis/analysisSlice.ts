/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { createSlice, type PayloadAction } from "@reduxjs/toolkit";
import { type AnalysisState, type Annotation } from "@/analysis/analysisTypes";
import { type UUID } from "@/core";
import type VarMap from "./analysisVisitors/varAnalysis/varMap";

const initialState: AnalysisState = {
  extensionAnnotations: {},
  knownVars: {},
};

const analysisSlice = createSlice({
  name: "analysis",
  initialState,
  reducers: {
    // TODO: if you see this action and it's still NOOP, remove it
    startAnalysis(
      state,
      action: PayloadAction<{ extensionId: UUID; analysisId: string }>
    ) {
      // NOOP
    },
    finishAnalysis(
      state,
      action: PayloadAction<{
        extensionId: UUID;
        analysisId: string;
        annotations: Annotation[];
      }>
    ) {
      const { extensionId, analysisId, annotations } = action.payload;

      // Clear out any existing annotations for this analysis.
      if (state.extensionAnnotations[extensionId]) {
        state.extensionAnnotations[extensionId] = state.extensionAnnotations[
          extensionId
        ].filter((x) => x.analysisId !== analysisId);
      } else {
        state.extensionAnnotations[extensionId] = [];
      }

      state.extensionAnnotations[extensionId] = [
        ...state.extensionAnnotations[extensionId],
        ...annotations,
      ];
    },
    setKnownVars(
      state,
      action: PayloadAction<{
        extensionId: UUID;
        vars: Map<string, VarMap>;
      }>
    ) {
      const { extensionId, vars } = action.payload;
      state.knownVars[extensionId] = vars;
    },
  },
});

export default analysisSlice;
