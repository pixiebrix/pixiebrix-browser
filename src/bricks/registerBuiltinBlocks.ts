/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import getAllEffects from "./effects/getAllEffects";
import getAllTransformers from "./transformers/getAllTransformers";
import getAllRenderers from "./renderers/getAllRenderers";
import getAllReaders, {
  registerReaderFactories,
} from "@/bricks/readers/getAllReaders";
import blockRegistry from "@/bricks/registry";

let registered = false;

function registerBuiltinBlocks() {
  if (registered) {
    console.warn(
      "registerBuiltinBlocks already called; multiple calls are unnecessary and may impact startup performance"
    );
  }

  blockRegistry.register([
    ...getAllTransformers(),
    ...getAllEffects(),
    ...getAllRenderers(),
    ...getAllReaders(),
  ]);

  registerReaderFactories();

  registered = true;
}

export default registerBuiltinBlocks;
