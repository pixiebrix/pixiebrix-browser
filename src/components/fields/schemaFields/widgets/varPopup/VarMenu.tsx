/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useEffect, useRef } from "react";
import styles from "./VarMenu.module.scss";

type VarMenuProps = {
  onClose?: () => void;
};

const VarMenu: React.FunctionComponent<VarMenuProps> = ({ onClose }) => {
  const rootElementRef = useRef<HTMLDivElement>(null);
  useEffect(() => {
    const handleClick = (event: MouseEvent) => {
      const parent = rootElementRef.current?.parentElement;
      if (parent && !parent.contains(event.target as Node)) {
        onClose?.();
      }
    };

    document.addEventListener("click", handleClick);
    return () => {
      document.removeEventListener("click", handleClick);
    };
  }, [onClose]);

  return (
    <div className={styles.menu} ref={rootElementRef}>
      <div className={styles.menuList}>
        All known variables will show up here
      </div>
    </div>
  );
};

export default VarMenu;
