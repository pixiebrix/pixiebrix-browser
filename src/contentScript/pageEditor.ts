/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { serializeError } from "serialize-error";
import {
  blockReducer,
  type IntermediateState,
  type ReduceOptions,
} from "@/runtime/reducePipeline";
import { type BlockConfig } from "@/blocks/types";
import { cloneDeep } from "lodash";
import ConsoleLogger from "@/utils/ConsoleLogger";
import { type SerializableResponse } from "@/pageScript/messenger/pigeon";
import apiVersionOptions from "@/runtime/apiVersionOptions";
import { $safeFind } from "@/helpers";
import { clearDynamicElements } from "@/contentScript/pageEditor/dynamic";
import { reactivateTab } from "./lifecycle";
import {
  BusinessError,
  CancelError,
  NoRendererError,
} from "@/errors/businessErrors";
import { uuidv4 } from "@/types/helpers";
import { type PanelPayload } from "@/sidebar/types";
import { HeadlessModeError } from "@/blocks/errors";
import { showTemporarySidebarPanel } from "@/contentScript/sidebarController";
import { stopInspectingNativeHandler } from "./pageEditor/elementPicker";
import { showModal } from "@/blocks/transformers/ephemeralForm/modalUtils";
import { createFrameSource } from "@/blocks/transformers/temporaryInfo/DisplayTemporaryInfo";
import { waitForTemporaryPanel } from "@/blocks/transformers/temporaryInfo/temporaryPanelProtocol";
import { type ApiVersion, type BlockArgsContext } from "@/types/runtimeTypes";
import { type UUID } from "@/types/stringTypes";
import { type RegistryId } from "@/types/registryTypes";

export type RunBlockArgs = {
  /**
   * The runtime API version to use
   */
  apiVersion: ApiVersion;
  /**
   * The IBlock configuration.
   */
  blockConfig: BlockConfig;
  /**
   * Context to render the BlockArg, should include @input, @options, and service context
   * @see makeServiceContext
   */
  context: BlockArgsContext;
  /**
   * Root jQuery selector to determine the root if the rootMode is "inherit".
   * @see BlockConfig.rootMode
   */
  rootSelector: string | undefined;
};

/**
 * Run a single block (e.g., for generating output previews)
 * @see BlockPreview
 */
export async function runBlock({
  blockConfig,
  context,
  apiVersion,
  rootSelector,
}: RunBlockArgs): Promise<unknown> {
  const versionOptions = apiVersionOptions(apiVersion);

  if (!versionOptions.explicitDataFlow) {
    throw new BusinessError(
      "Preview only supported for mods using runtime v2 or later"
    );
  }

  const state: IntermediateState = {
    context,
    // Can pick any index. It's only used for adding context to log messages, and we're disabling value logging
    // below with `logValues: false`
    index: 0,
    // Force isLastBlock so blockReducer does not complain about the outputKey being forced to undefined
    isLastBlock: true,
    // `root` is over-ridden below if rootSelector is provided
    root: document,
    // We're forcing apiVersion: 2 or higher above values must come from the context
    previousOutput: {},
  };

  if (rootSelector) {
    // Handle non-document contexts. If the selector is unique, this gives the root that would be available at runtime.
    // Differences in behavior:
    // - For triggers, the PixieBrix looks for the closest ancestor to the DOM event target matching the selector
    //   See TriggerExtensionPoint.eventHandler for reference
    // - For multi-menus (not currently available in the Page Editor), the below logic returns an arbitrary menu
    const rootElement = $safeFind(rootSelector);
    if (rootElement.length > 0) {
      state.root = rootElement.get(0);
    }
  }

  const options: ReduceOptions = {
    ...versionOptions,
    branches: [],
    headless: true,
    logValues: false,
    logger: new ConsoleLogger(),
    // Excluding runId will prevent the run from being stored in traces
    runId: null,
    extensionId: null,
  };

  // Exclude the outputKey so that `output` is the output of the brick. Alternatively we could have taken then
  // value from the context[outputKey] from the return value of blockReducer
  const { output } = await blockReducer(
    { ...blockConfig, outputKey: undefined },
    state,
    options
  );

  return cloneDeep(output) as SerializableResponse;
}

type Location = "modal" | "panel";

/**
 * Run a single renderer (e.g. - for running a block preview)
 *
 * Renderers need to be run with try-catch, catch the HeadlessModeError, and
 * use that to send the panel payload to the sidebar (or other target)
 * @see SidebarExtensionPoint
 *  starting on line 184, the call to reduceExtensionPipeline(),
 *  wrapped in a try-catch
 * @see executeBlockWithValidatedProps
 *  starting on line 323, the runRendererPipeline() function
 *
 * Note: Currently only implemented for the temporary sidebar panels
 * @see useDocumentPreviewRunBlock
 */
export async function runRendererBlock({
  extensionId,
  blueprintId,
  runId,
  title,
  args,
  location,
}: {
  extensionId: UUID;
  blueprintId: RegistryId | null;
  runId: UUID;
  title: string;
  args: RunBlockArgs;
  location: Location;
}): Promise<void> {
  const nonce = uuidv4();

  let payload: PanelPayload;
  try {
    await runBlock(args);
    // We're expecting a HeadlessModeError (or other error) to be thrown in the line above
    // noinspection ExceptionCaughtLocallyJS
    throw new NoRendererError();
  } catch (error) {
    if (error instanceof HeadlessModeError) {
      payload = {
        key: nonce,
        blockId: error.blockId,
        args: error.args,
        ctxt: error.ctxt,
        extensionId,
        runId,
      };
    } else {
      payload = {
        key: nonce,
        error: serializeError(error),
        extensionId,
        runId,
      };
    }

    if (location === "panel") {
      showTemporarySidebarPanel({
        // Pass extension id so previous run is cancelled
        extensionId,
        blueprintId,
        nonce,
        heading: title,
        payload,
      });
    } else if (location === "modal") {
      const controller = new AbortController();
      const url = await createFrameSource(nonce, "modal");

      showModal({ url, controller });

      try {
        await waitForTemporaryPanel(nonce, {
          extensionId,
          blueprintId,
          nonce,
          heading: title,
          payload,
        });
      } catch (error) {
        // Match behavior of Display Temporary Info
        if (error instanceof CancelError) {
          // NOP
        } else {
          throw error;
        }
      } finally {
        controller.abort();
      }
    } else {
      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions -- dynamic check for never
      throw new Error(`Support for previewing in ${location} not implemented`);
    }
  }
}

export async function resetTab(): Promise<void> {
  stopInspectingNativeHandler();
  await clearDynamicElements({});
  await reactivateTab();
}
