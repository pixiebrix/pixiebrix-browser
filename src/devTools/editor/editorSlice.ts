/*
 * Copyright (C) 2020 Pixie Brix, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { IconConfig, Metadata, Schema, ServiceDependency } from "@/core";
import { ElementInfo } from "@/nativeEditor/frameworks";
import { MenuPosition } from "@/extensionPoints/menuItemExtension";
import { MenuDefaultOptions as ContextMenuDefaultOptions } from "@/extensionPoints/contextMenu";
import { BlockPipeline } from "@/blocks/combinators";
import { Trigger } from "@/extensionPoints/triggerExtension";
import { Menus } from "webextension-polyfill-ts";

export interface ReaderReferenceFormState {
  metadata: Metadata;
}

export interface ReaderFormState {
  _new?: boolean;
  metadata: Metadata;
  outputSchema: Schema;
  definition: {
    /**
     * Reader type corresponding to built-in reader factory, e.g., jquery, react.
     */
    type: string | null;
    selector: string | null;
    selectors: { [field: string]: string };
    optional: boolean;
  };
}

export function isCustomReader(
  reader: ReaderFormState | ReaderReferenceFormState
): reader is ReaderFormState {
  return "definition" in reader;
}

export type ElementType =
  | "menuItem"
  | "trigger"
  | "panel"
  | "contextMenu"
  | "actionPanel";

export interface BaseFormState {
  readonly uuid: string;
  readonly type: ElementType;

  installed?: boolean;
  autoReload?: boolean;

  label: string;

  services: ServiceDependency[];

  readers: (ReaderFormState | ReaderReferenceFormState)[];

  extensionPoint: unknown;

  extension: unknown;
}

export interface ContextMenuFormState extends BaseFormState {
  type: "contextMenu";

  extensionPoint: {
    metadata: Metadata;
    definition: {
      defaultOptions: ContextMenuDefaultOptions;
      documentUrlPatterns: string[];
      contexts: Menus.ContextType[];
      isAvailable: {
        matchPatterns: string;
        selectors: string;
      };
    };
  };

  extension: {
    title: string;
    action: BlockPipeline;
  };
}

export interface TriggerFormState extends BaseFormState {
  type: "trigger";

  extensionPoint: {
    metadata: Metadata;
    definition: {
      rootSelector: string | null;
      trigger: Trigger;
      isAvailable: {
        matchPatterns: string;
        selectors: string;
      };
    };
  };

  extension: {
    action: BlockPipeline;
  };
}

export interface ActionPanelFormState extends BaseFormState {
  type: "actionPanel";

  extensionPoint: {
    metadata: Metadata;
    definition: {
      isAvailable: {
        matchPatterns: string;
        selectors: string;
      };
    };
  };

  extension: {
    heading: string;
    body: BlockPipeline;
  };
}

export type PanelTraits = {
  style: {
    mode: "default" | "inherit";
  };
};

export interface PanelFormState extends BaseFormState {
  type: "panel";

  containerInfo: ElementInfo;

  extensionPoint: {
    metadata: Metadata;
    definition: {
      containerSelector: string;
      position?: MenuPosition;
      template: string;
      isAvailable: {
        matchPatterns: string;
        selectors: string;
      };
    };
    traits: PanelTraits;
  };

  extension: {
    heading: string;
    body: BlockPipeline;
    collapsible?: boolean;
    shadowDOM?: boolean;
  };
}

export interface ActionFormState extends BaseFormState {
  type: "menuItem";

  containerInfo: ElementInfo;

  extensionPoint: {
    metadata: Metadata;
    definition: {
      containerSelector: string;
      position?: MenuPosition;
      template: string;
      isAvailable: {
        matchPatterns: string;
        selectors: string;
      };
    };
    traits?: {
      style: {
        mode: "default" | "inherit";
      };
    };
  };

  extension: {
    caption: string;
    dynamicCaption?: boolean;
    icon?: IconConfig;
    action: BlockPipeline;
  };
}

export type FormState =
  | ActionFormState
  | ActionPanelFormState
  | TriggerFormState
  | PanelFormState
  | ContextMenuFormState;

export interface EditorState {
  selectionSeq: number;
  inserting: ElementType | null;
  activeElement: string | null;
  error: string | null;
  dirty: Record<string, boolean>;
  knownEditable: string[];
  readonly elements: FormState[];

  /**
   * True if error is because user does not have access to beta features
   */
  beta?: boolean;
}

export const initialState: EditorState = {
  selectionSeq: 0,
  activeElement: null,
  error: null,
  beta: false,
  elements: [],
  knownEditable: [],
  dirty: {},
  inserting: null,
};

export const editorSlice = createSlice({
  name: "editor",
  initialState,
  reducers: {
    toggleInsert: (state, action: PayloadAction<ElementType>) => {
      state.inserting = action.payload;
      state.beta = false;
      state.error = null;
    },
    markEditable: (state, action: PayloadAction<string>) => {
      state.knownEditable.push(action.payload);
    },
    addElement: (state, action: PayloadAction<FormState>) => {
      const element = action.payload;
      state.elements.push(element);
      state.error = null;
      state.beta = false;
      state.activeElement = element.uuid;
      state.selectionSeq++;
    },
    betaError: (state, action: PayloadAction<{ error: string }>) => {
      state.error = action.payload.error;
      state.beta = true;
      state.activeElement = null;
    },
    adapterError: (
      state,
      action: PayloadAction<{ uuid: string; error: unknown }>
    ) => {
      const { uuid, error } = action.payload;
      if (error instanceof Error) {
        state.error = error.message ?? "Unknown error";
      } else {
        state.error = error.toString() ?? "Unknown error";
      }
      state.beta = false;
      state.activeElement = uuid;
      state.selectionSeq++;
    },
    selectInstalled: (state, actions: PayloadAction<FormState>) => {
      const index = state.elements.findIndex(
        (x) => x.uuid === actions.payload.uuid
      );
      if (index >= 0) {
        // safe because we're getting it from findIndex
        // eslint-disable-next-line security/detect-object-injection
        state.elements[index] = actions.payload;
      } else {
        state.elements.push(actions.payload);
      }
      state.error = null;
      state.beta = null;
      state.activeElement = actions.payload.uuid;
      state.selectionSeq++;
    },
    resetInstalled: (state, actions: PayloadAction<FormState>) => {
      const { uuid } = actions.payload;
      const index = state.elements.findIndex((x) => x.uuid === uuid);
      if (index >= 0) {
        // safe because we're getting it from findIndex
        // eslint-disable-next-line security/detect-object-injection
        state.elements[index] = actions.payload;
      } else {
        state.elements.push(actions.payload);
      }
      state.dirty[uuid] = false;
      state.error = null;
      state.beta = null;
      state.activeElement = uuid;
      state.selectionSeq++;
    },
    selectElement: (state, action: PayloadAction<string>) => {
      if (!state.elements.some((x) => action.payload === x.uuid)) {
        throw new Error(`Unknown dynamic element: ${action.payload}`);
      }
      state.error = null;
      state.beta = null;
      state.activeElement = action.payload;
      state.selectionSeq++;
    },
    markSaved: (state, action: PayloadAction<string>) => {
      const element = state.elements.find((x) => action.payload === x.uuid);
      if (!element) {
        throw new Error(`Unknown dynamic element: ${action.payload}`);
      }
      if (!element.installed) {
        state.knownEditable.push(
          element.extensionPoint.metadata.id,
          ...element.readers
            .filter((x) => isCustomReader(x))
            .map((x) => x.metadata.id)
        );
      }

      for (const reader of element.readers) {
        if (isCustomReader(reader)) {
          reader._new = false;
        }
      }

      element.installed = true;
      state.dirty[element.uuid] = false;
      // force a reload so the _new flags are correct on the readers
      state.selectionSeq++;
    },
    // sync the redux state with the form state
    updateElement: (state, action: PayloadAction<FormState>) => {
      const { uuid } = action.payload;
      const index = state.elements.findIndex((x) => x.uuid === uuid);
      if (index < 0) {
        throw new Error(`Unknown dynamic element: ${uuid}`);
      }
      // safe b/c generated from findIndex
      // eslint-disable-next-line security/detect-object-injection
      state.elements[index] = action.payload;
      state.dirty[uuid] = true;
    },
    removeElement: (state, action: PayloadAction<string>) => {
      const uuid = action.payload;
      if (state.activeElement === uuid) {
        state.activeElement = null;
      }
      state.elements.splice(
        state.elements.findIndex((x) => x.uuid === uuid),
        1
      );
      delete state.dirty[uuid];
    },
  },
});

export const actions = editorSlice.actions;
