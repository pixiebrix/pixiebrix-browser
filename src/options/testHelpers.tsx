/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import settingsSlice from "@/store/settingsSlice";
import { blueprintModalsSlice } from "@/options/pages/blueprints/modals/blueprintModalsSlice";
import { createRenderWithWrappers } from "@/testUtils/testHelpers";
import blueprintsSlice from "@/options/pages/blueprints/blueprintsSlice";
import { recipesSlice } from "@/recipes/recipesSlice";

const renderWithWrappers = createRenderWithWrappers(() =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      settings: settingsSlice.reducer,
      options: extensionsSlice.reducer,
      blueprintModals: blueprintModalsSlice.reducer,
      blueprints: blueprintsSlice.reducer,
      recipes: recipesSlice.reducer,
    },
  })
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
