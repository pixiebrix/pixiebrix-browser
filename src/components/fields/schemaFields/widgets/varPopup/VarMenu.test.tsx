/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { render } from "@/pageEditor/testHelpers";
import React, { type MutableRefObject } from "react";
import VarMenu from "@/components/fields/schemaFields/widgets/varPopup/VarMenu";
import { formStateFactory } from "@/testUtils/factories/pageEditorFactories";
import { actions as editorActions } from "@/pageEditor/slices/editorSlice";
import { waitForEffect } from "@/testUtils/testHelpers";
import registerBuiltinBlocks from "@/blocks/registerBuiltinBlocks";
import analysisSlice from "@/analysis/analysisSlice";
import VarAnalysis from "@/analysis/analysisVisitors/varAnalysis/varAnalysis";

const TestWrapper: React.FunctionComponent<{
  renderMenu: (
    inputElementRef: MutableRefObject<HTMLInputElement>
  ) => React.ReactNode;
}> = ({ renderMenu }) => {
  const inputElementRef = React.useRef<HTMLInputElement>(null);
  return (
    <>
      <input ref={inputElementRef} />
      {renderMenu(inputElementRef)}
    </>
  );
};

beforeAll(() => {
  registerBuiltinBlocks();
});

describe("VarMenu", () => {
  test("shows variables not available message", async () => {
    const formState = formStateFactory();

    const wrapper = render(
      <TestWrapper
        renderMenu={(inputElementRef) => (
          <VarMenu
            inputElementRef={inputElementRef}
            onClose={jest.fn()}
            onVarSelect={jest.fn()}
            likelyVariable="@foo"
          />
        )}
      />,
      {
        setupRedux(dispatch) {
          dispatch(editorActions.addElement(formState));
          dispatch(editorActions.selectElement(formState.uuid));
        },
      }
    );

    await waitForEffect();

    expect(
      wrapper.getByText("Available variables have not been computed yet.")
    ).toBeInTheDocument();
  });

  test("shows view for unmatched top-level variable", async () => {
    const formState = formStateFactory();

    const wrapper = render(
      <TestWrapper
        renderMenu={(inputElementRef) => (
          <VarMenu
            inputElementRef={inputElementRef}
            onClose={jest.fn()}
            onVarSelect={jest.fn()}
            likelyVariable="@foo"
          />
        )}
      />,
      {
        async setupRedux(dispatch) {
          dispatch(editorActions.addElement(formState));
          dispatch(editorActions.selectElement(formState.uuid));
          dispatch(
            editorActions.setElementActiveNodeId(
              formState.extension.blockPipeline[0].instanceId
            )
          );

          // Run analysis directly
          const analysis = new VarAnalysis({ trace: [], modState: {} });
          await analysis.run(formState);

          dispatch(
            analysisSlice.actions.setKnownVars({
              extensionId: formState.uuid,
              vars: analysis.getKnownVars(),
            })
          );
        },
      }
    );

    await waitForEffect();

    // Can't include @foo in query because it's within a span
    expect(wrapper.queryByText("No variables found for")).toBeInTheDocument();
  });

  test("shows view for match", async () => {
    const formState = formStateFactory();

    const wrapper = render(
      <TestWrapper
        renderMenu={(inputElementRef) => (
          <VarMenu
            inputElementRef={inputElementRef}
            onClose={jest.fn()}
            onVarSelect={jest.fn()}
            likelyVariable="@inp"
          />
        )}
      />,
      {
        async setupRedux(dispatch) {
          dispatch(editorActions.addElement(formState));
          dispatch(editorActions.selectElement(formState.uuid));
          dispatch(
            editorActions.setElementActiveNodeId(
              formState.extension.blockPipeline[0].instanceId
            )
          );

          // Run analysis directly
          const analysis = new VarAnalysis({ trace: [], modState: {} });
          await analysis.run(formState);

          dispatch(
            analysisSlice.actions.setKnownVars({
              extensionId: formState.uuid,
              vars: analysis.getKnownVars(),
            })
          );
        },
      }
    );

    await waitForEffect();

    expect(
      wrapper.queryByText("No variables found for")
    ).not.toBeInTheDocument();

    // @inp matches @input
    expect(wrapper.queryByText("@input")).toBeInTheDocument();

    expect(wrapper.asFragment()).toMatchSnapshot();
  });
});
