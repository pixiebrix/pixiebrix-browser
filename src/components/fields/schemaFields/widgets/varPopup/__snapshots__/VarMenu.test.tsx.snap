// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VarMenu shows view for match 1`] = `
<DocumentFragment>
  <input />
  <div
    class="menu"
    style="max-height: -16px; transform: translate3d(0, 8px, 0);"
  >
    <div
      class="body"
    >
      <div
        class="sourceItem"
      >
        <div
          class="label"
        >
          Starter Brick: Panel
        </div>
        <ul
          style="border: 0px; padding: 0px; margin: 0.5em 0px 0.5em 0.125em; list-style: none; background-color: rgb(253, 255, 255);"
        >
          <li
            style="padding: 0px; margin: 0px;"
          >
            <ul
              style="padding: 0px; margin: 0px; list-style: none;"
            >
              <li
                class="active"
                style="position: relative; padding-top: 0.25em; margin-left: 0px; padding-left: 0px; display: flex; align-items: center; flex-wrap: wrap;"
              >
                <div
                  style="display: inline-block; cursor: pointer; padding: 4px; margin-right: 10px; background-color: rgb(240, 239, 242); border-radius: 2px;"
                >
                  <div
                    style="margin-left: 0px; transition: 150ms; transform: rotateZ(0deg); transform-origin: 45% 50%; position: relative; line-height: 1.1em; font-size: 0.75em; height: 12px; width: 12px; display: flex; justify-content: center; align-items: center; color: rgb(46, 36, 65);"
                  >
                    â–¶
                  </div>
                </div>
                <label
                  style="display: inline-block; color: rgb(46, 36, 65); word-break: initial; text-indent: -0.5em; margin: 0px; padding: 0px; cursor: pointer;"
                >
                  <button
                    class="btn"
                    type="button"
                  >
                    @input
                  </button>
                </label>
                <span
                  style="padding-left: 0.5em; cursor: default; color: rgb(64, 175, 108);"
                >
                  <span>
                    <span
                      style="margin-left: 0.3em; margin-right: 0.3em;"
                    >
                      {}
                    </span>
                     4 keys
                  </span>
                </span>
                <ul
                  style="padding: 0px; margin: 0px; list-style: none; display: block; width: 100%;"
                />
              </li>
            </ul>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="footer"
    >
      Use up/down keys to navigate, tab to select
    </div>
  </div>
</DocumentFragment>
`;
